### PR TITLE
Gui cleanup4

### PR DESCRIPTION
**Pull request by alexvins**

### CODE REVIEW

**Comment by karliss** [comment]
Aren't shadow parts https://github.com/vcmi/vcmi/blob/5c09f751b343fc3d46dd5c3f4f08b0614cf3522f/client/windows/CWindowObject.cpp#L223  that were changed in GUI cleanup 3 leaking memory  due to `shadow.defActions` not having DISPOSE flag?

**Comment by ArseniyShestakov** [comment]
@karliss I'm sorry if I misunderstand your point, but all these new branches you see have end goal to get rid of ```DISPOSE``` completely. Though since we change them step-by-step there will be some leaks until refactoring of all UI is complete.

**Comment by karliss** [comment]
No problem, I wasn't sure how clean you you are trying to have the develop branch during this refactoring.

**Comment by ArseniyShestakov** [comment]
@karliss
Regardless thanks for feedback. We usually trying to keep develop as stable as possible so comments like that are really appreciated!

Unfortunately in this particular case we decide to land code as soon as possible since refactoring of UI require huge amount of work and it's easier to test each group of changes separately rather than trying to rewrite it all within single branch.

**Comment by ArseniyShestakov** [comment]
Ok there is regression: when on Linux window minimized and then opened again renderer is recreated.
After that cursor become invisible.

**Comment by alexvins** [comment]
@karliss it should not leak

there is `OBJECT_CONSTRUCTION_CUSTOM_CAPTURING(255);` just before that line

**Comment by karliss** [comment]
@alexvins CIntObject.cpp destructor checks DISPOSE flag in parents defActions and childeren recActions. Having it only for children is insufficient.

**Comment by alexvins** [comment]
@karliss shadow is default constructed `CIntObject` it should have defActions = 255

**Comment by karliss** [comment]
Just confirmed it with a debugger `shadow->defactions == 239` most likely due to https://github.com/vcmi/vcmi/blob/develop/client/gui/CIntObject.cpp#L32 and https://github.com/vcmi/vcmi/blob/develop/client/windows/CWindowObject.cpp#L216 .

**Comment by alexvins** [comment]
OK, fixed

File: client/battle/CBattleInterface.h [hunk 1]
```diff
@@ -121,14 +121,26 @@ class CBattleInterface : public CIntObject
 private:
 	SDL_Surface *background, *menu, *amountNormal, *amountNegative, *amountPositive, *amountEffNeutral, *cellBorders, *backgroundWithHexes;
 
-	CButton *bOptions, *bSurrender, *bFlee, *bAutofight, *bSpell,
-		* bWait, *bDefence, *bConsoleUp, *bConsoleDown, *btactNext, *btactEnd;
-	CBattleConsole *console;
-	CBattleHero *attackingHero, *defendingHero; //fighting heroes
-	CStackQueue *queue;
+	std::shared_ptr<CButton> bOptions;
+	std::shared_ptr<CButton> bSurrender;
```
Comment (FeniksFire) [inline client/battle/CBattleInterface.h]: these buttons, shouldn't be unique_ptr?

Comment (ArseniyShestakov) [reply client/battle/CBattleInterface.h]: @FeniksFire why?

Usage of ```unique_ptr``` apply certain requirements on how headers are organized too and ours aren't very good. You know...

Comment (FeniksFire) [reply client/battle/CBattleInterface.h]: @ArseniyShestakov and why you need shared_ptr? From first point of view something like "button" shouldn't be shared, and even modified. So sole owner for buttons sounds good.

Comment (ArseniyShestakov) [reply client/battle/CBattleInterface.h]: It does make sense, but I suppose at least for now we slap ```shared``` on everything so we can keep internals of GH working until we actually rewrite the whole thing. After all GH itself not yet use smart pointers I suppose.

Comment (alexvins) [reply client/battle/CBattleInterface.h]: All widgets must be `shared_ptr`, later child links in CIntObjects to be changed to shared_ptr and parent link to weak_ptr

Comment (alexvins) [reply client/battle/CBattleInterface.h]: @FeniksFire  What is incorrectly written?

**Comment by ArseniyShestakov** [comment]
Ok so I finally able to get my hands on branch. Regression found:

- If you start game with multiple towns and switch between them inside town screen then when you attempt to leave screen you'll have to click on button multiple times since all screens remain alive in background. This also cause crashes.

**Comment by dydzio0614** [comment]
Little testing revealed 1 bug so far - cursor type does not change in some cases where it should. When interacting with adventure map object to make dialog window appear, the cursor does not change to arrow, but stays the same type as when clicking on object to pick it. Same with starting battle by attacking and sword icon being active until actual battle starts, etc.

Also some crash happened that may be unrelated to the branch. At townscreen crash happened. At 1st I was clicking various buildings, including marketplace and mage guild, then I was moving cursor over empty creature slots. At that moment crash occured, probably when cursor was at border between upper and lower slots

**Comment by ArseniyShestakov** [comment]
Ok I tested it again today and it's seems alright to merge.
Hopefully I didn't break something. :-)

**Pull request merged**